In [1]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import loadmat

mnist = loadmat('datasets/mnist-original.mat')
X = mnist['data']
y = mnist['label']
X = X.T
y = y.reshape(70000,)

In [2]:
from sklearn.decomposition import PCA
pca = PCA(0.90)
pca.fit(X)
X_reduction = pca.transform(X)

In [3]:
pca.n_components_

87

In [21]:
from sklearn.ensemble import RandomForestClassifier

rf_clf = RandomForestClassifier(n_estimators=1000, max_leaf_nodes=5000, oob_score=True, random_state=666, n_jobs=-1)
rf_clf.fit(X_reduction, y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=5000,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=-1,
            oob_score=True, random_state=666, verbose=0, warm_start=False)

In [22]:
rf_clf.oob_score_

0.9547

### n_estimators=1000, max_leaf_nodes=5000的意思是 使用了1000个决策树的分类器，而每一个分类器最多只能有5000个叶子节点

### 这段代码中，使用的取数据的方法是老师课上介绍的自助法，oob_score_的结果0.9547是对从来没有取到过的数据进行预测所得到的准确率

### 代码中使用的测试数据集是mnist手写数据集，其官方网站是http://yann.lecun.com/exdb/mnist/